### ISSUE TITLE
monitord can't insert monitoring data 

### ISSUE THREAD

**Issue by atodorov-storpool** [OpenNebula/one#5196]
**Description**

Having the following in monitord.log between the successful :

```
...
Fri Dec  4 18:17:22 2020 [Z0][HMM][I]: Successfully monitored VM: 7
Fri Dec  4 18:17:22 2020 [Z0][HMM][I]: Successfully monitored VM: 8
Fri Dec  4 18:17:22 2020 [Z0][HMM][I]: Successfully monitored VM: 5
Fri Dec  4 18:17:22 2020 [Z0][HMM][I]: Successfully monitored VM: 3
Fri Dec  4 18:17:22 2020 [Z0][HMM][I]: Successfully monitored VM: 9
Fri Dec  4 18:17:22 2020 [Z0][HMM][I]: Successfully monitored VM: 12
Fri Dec  4 18:17:26 2020 [Z0][ONE][E]: SQL command was: REPLACE INTO vm_monitoring (vmid, last_poll, body) VALUES (4,1607098646,'<MONITORING><CPU><![CDATA[30.78]]></CPU><DISKRDBYTES><![CDATA[796155776]]></DISKRDBYTES><DISKRDIOPS><![CDATA[27270]]></DISKRDIOPS><DISKWRBYTES><![CDATA[5414258176]]></DISKWRBYTES><DISKWRIOPS><![CDATA[883527]]></DISKWRIOPS><ID><![CDATA[4]]></ID><MEMORY><![CDATA[16849780]]></MEMORY><NETRX><![CDATA[15289195964]]></NETRX><NETTX><![CDATA[31285031216]]></NETTX><TIMESTAMP><![CDATA[1607098646]]></TIMESTAMP></MONITORING>'), error 1046 : No database selected
Fri Dec  4 18:17:26 2020 [Z0][HMM][E]: Unable to write monitoring to DB
Fri Dec  4 18:17:26 2020 [Z0][HMM][I]: Successfully monitored VM: 5
Fri Dec  4 18:17:26 2020 [Z0][HMM][I]: Successfully monitored VM: 6
Fri Dec  4 18:17:26 2020 [Z0][HMM][I]: Successfully monitored VM: 7
Fri Dec  4 18:17:26 2020 [Z0][ONE][E]: SQL command was: SELECT vm_monitoring.body FROM vm_monitoring WHERE vmid = 9 AND last_poll=(SELECT MAX(last_poll) FROM vm_monitoring WHERE vmid = 9), error 1046 : No database selected
...
```

**To Reproduce**
Somehow force reconnect on a single MySQL connection should observe something like above.

**Expected behavior**
The SQL commands should succeed

**Details**
 - Affected Component: [Core]
 - Hypervisor: [KVM]
 - Version: [Let's say development ;)]

**Additional context**
Looking at the code, the reconnect is set when creating the MySQL connections:
https://github.com/OpenNebula/one/blob/master/src/sql/MySqlDB.cc#L206
Then the connection is made without setting the database name:
https://github.com/OpenNebula/one/blob/master/src/sql/MySqlDB.cc#L209
And after that the database is selected with a SQL query:
https://github.com/OpenNebula/one/blob/master/src/sql/MySqlDB.cc#L251

My wild guess is that after reconnecting the database is indeed not selected...

Following https://dev.mysql.com/doc/c-api/5.7/en/c-api-auto-reconnect.html
there is a recommendation tomonitor the _mysql_thread_id_  and the USE statement re-issued when changed( or use the database _mysql_real_connect()_ but then _mysql_set_character_set()_ should be handled too?)

## Progress Status
- [ ] Branch created 
- [ ] Code committed to development branch
- [ ] Testing - QA
- [ ] Documentation 
- [ ] Release notes - resolved issues, compatibility, known issues
- [ ] Code committed to upstream release/hotfix branches
- [ ] Documentation committed to upstream release/hotfix branches

**Comment by atodorov-storpool** [comment]
Pull request: https://github.com/OpenNebula/one/pull/5198

**Comment by paczerny** [comment]
Hi Anton,
I wasn't able to reproduce the issue. I never get the error 'No database selected'.
I tested it only by restarting mysql server, after the restart everything works fine.

Do you have any hint how to force reconnect on single connection?

**Comment by paczerny** [comment]
In the mysql client I can do 'show processlist;' and 'kill id', to close only one connection.
But still I don't get the 'No database selected' error

**Comment by atodorov-storpool** [comment]
This happened only once during restart, and probably a weird race condition between MySQL and oned was involved.

> In the mysql client I can do 'show processlist;' and 'kill id', to close only one connection.
> But still I don't get the 'No database selected' error

How the connection pool is utilized? By round-robin? Is it possible that the broken connection was not used? In the MySQL docs there is hint to check the thread ID to detect is the connection re-created or not. Maybe if there is debug log to show the connection thread ids it could be detected is the "reconnected" connection USD or not? or reduce the number of Mysql connections?

**Comment by paczerny** [comment]
It uses FIFO queue. I killed a lot of connections and I waited long enough to make sure the connection was reused

**Comment by paczerny** [comment]
Your commit is merged, closing the issue. Although I can't test that it solves the issue

**Comment by atodorov-storpool** [comment]
It is clear that it is fixed in master(6.0), just mentioning that it is happening with 5.12.7-ee too:

> Tue Jan 19 18:30:22 2021 [Z0][ONE][E]: SQL command was: SELECT vm_monitoring.body FROM vm_monitoring WHERE vmid = 369 AND last_poll=(SELECT MAX(last_poll) FROM vm_monitoring WHERE vmid = 369), error 1046 : No database selected
> Tue Jan 19 18:30:22 2021 [Z0][HMM][I]: Successfully monitored VM: 369
> Tue Jan 19 18:30:22 2021 [Z0][HMM][I]: Successfully monitored VM: 372
> Tue Jan 19 18:30:22 2021 [Z0][HMM][I]: Successfully monitored VM: 374
> Tue Jan 19 18:30:22 2021 [Z0][HMM][I]: Successfully monitored VM: 378
> Tue Jan 19 18:30:22 2021 [Z0][HMM][I]: Successfully monitored VM: 386
> Tue Jan 19 18:30:22 2021 [Z0][HMM][I]: Successfully monitored VM: 388
> Tue Jan 19 18:30:22 2021 [Z0][HMM][I]: Successfully monitored VM: 389
> Tue Jan 19 18:30:22 2021 [Z0][ONE][E]: SQL command was: REPLACE INTO vm_monitoring (vmid, last_poll, body) VALUES (391,1611055822,'<MONITORING><CPU><![CDATA[3.04]]></CPU><DISKRDBYTES><![CDATA[931160064]]></DISKRDBYTES><DISKRDIOPS><![CDATA[36110]]></DISKR
> DIOPS><DISKWRBYTES><![CDATA[5243668992]]></DISKWRBYTES><DISKWRIOPS><![CDATA[286468]]></DISKWRIOPS><DISK_SIZE><ID><![CDATA[0]]></ID><SIZE><![CDATA[9583]]></SIZE></DISK_SIZE><ID><![CDATA[391]]></ID><MEMORY><![CDATA[2139124]]></MEMORY><MONITORING><CPU><![CD
> ATA[0.0]]></CPU><DISKRDBYTES><![CDATA[8192]]></DISKRDBYTES><DISKRDIOPS><![CDATA[1]]></DISKRDIOPS><DISKWRBYTES><![CDATA[16957952]]></DISKWRBYTES><DISKWRIOPS><![CDATA[2454]]></DISKWRIOPS><DISK_SIZE><![CDATA[\n      ]]></DISK_SIZE><MEMORY><![CDATA[2087364]]
> ></MEMORY><NETRX><![CDATA[851767]]></NETRX><NETTX><![CDATA[35425]]></NETTX><NIC_STATS><![CDATA[\n      ]]></NIC_STATS><STATE><![CDATA[a]]></STATE></MONITORING><NETRX><![CDATA[789964525]]></NETRX><NETTX><![CDATA[12491355802]]></NETTX><NIC_STATS><ID><![CDA
> TA[0]]></ID><RX><![CDATA[789966116]]></RX><TX><![CDATA[12491387525]]></TX></NIC_STATS><TIMESTAMP><![CDATA[1611055822]]></TIMESTAMP></MONITORING>'), error 1046 : No database selected
>
